whctools/views.py

```diff
@@ -401,98 +407,107 @@ def list_acl_members(request, acl_pk=""):
-                        "character_id": m.id,
-                    }
-                    for m in character["alts"]
-                    if m.character_name != character["main"]["name"]
-                ],
-                key=lambda x: x["name"],
-            )
-        )
-
-        acl_alt_names = [alt["name"] for alt in character["alts"]]
-
-        character["complete_alts"] = list(
-            sorted(
-                [
-                    {
-                        "name": m.character_name,
-                        "corp": m.corporation_name,
-                        "alliance": m.alliance_name,
-                        "portrait_url": m.portrait_url(32),
-                    }
-                    for m in character["complete_alts"]
-                    if m.character_name not in acl_alt_names
-                    and m.character_name != character["main"]["name"]
-                ],
-                key=lambda x: x["name"],
-            )
-        )
-
-    # Pre-compute aggregates
-    total_mains = len(alphabetical_mains)
-    total_chars = acl_obj.characters.count()
+    mains_set = set([]) # just mains in ACL
+    players_set = set([]) # includes mains in and not in ACL
+    for member in members_on_acl:
+        name = member.character_name
+        char_id = member.character_id
+        main = None
+        corp = member.corporation_name
+        alliance = member.alliance_name
+        error = None
+
+        main_character = get_main_character_from_evecharacter(member)
+        if main_character is None:
+            main = '?'
+            error = 'Orphaned character'
+            logger.info(f"WHC ACL '{acl_pk}': character '{name}' is an orphan with no main")
+        else:
+            main = main_character.character_name
+            players_set.add(main)
+
+        if name == main:
+            mains_set.add(main)
+
+        if not is_character_in_allowed_corp(member):
+            logger.info(f"WHC ACL '{acl_pk}': character '{name}' is in an invalid corp or alliance")
+            error = 'Disallowed corp/alliance'
+        # I believe this isn't possible, since a user has to swap Uni mains in
+        # order for them to remain in the Uni.
+        if main_character is not None and not is_character_in_allowed_corp(main_character):
+            logger.info(f"WHC ACL '{acl_pk}': character '{name}' has a main '{main}' in an invalid corp or alliance")
+            error = 'Main in disallowed corp/alliance'
+
+        char_list.append({
+            "name": name,
+            "id": char_id,
+            "main": main,
+            "corp": corp,
+            "alliance": alliance,
+            "error": error,
+            "is_main": (name==main),
+            "main_in_acl": False, # We'll backfill this
+        })
+
+    # Backfill mains
+    for char in char_list:
+        if char["main"] in mains_set:
+            char["main_in_acl"] = True
+
+    # sorted() Key function
+    class ACLSorter(object):
```

Comment:
is it worth moving it to its own class outside of the function, so it can be used anywhere we want to sort the ACL from, not just in here?

Reply:
We could, but it's not actually sorting the ACL. It's sorting the locally-synthesized data structure behind `char_list` which is unique to this function, and it's not anywhere near generic enough to really be wrangled for anything else. Hypothetically, someone could export a the `sorted_char_list` from `list_acl_members` somewhere, mangle it, than want to re-sort it, but it'd be pretty gnarly since they'd be mis-using a django view for something unrelated. Nobody should really be importing from `views.py`, so the generic way of writing this would be to define a reusable row-wise data structure for `char_list` somewhere else (`utils.py` probably) and then toss this Key function alongside it. Probably would need to hoist the list generator outside, too, since it's useless without it.

The big downside to all of that is it makes it a lot easier to make silly mistakes when changing the field included in a `char_list` row. I.e., if you added a new field or removed one, then it's easy to overlook the sort needs to get updated in addition to the template (which is already over yonder in `templates/whctools/list_acl_members.html`).

Some of the ugliness is because Python 3 doesn't like old-style `cmp` arguments to `sorted` anymore. I could've written this as a `cmp`-derivative and then used `functools.cmp_to_key`, but that mostly just changes the style. You'd have an `char_cmp` function which would do basically the same as `__lt__` but returning -1,0,1 instead of T/F. But you'd still have the same aspect of having an embedded function here, since writing this as a lambda would be horrifically unreadable.

If you think it looks stylistically better or more readable to use a scoped `cmp`-style function instead of a scoped class masquerading as a Key function, I'm happy write it that way instead. That's not too hard. I'm still probably leaning towards having it local to `list_acl_members` to avoid mistakes when changing it, though.